contracts/instance/InstanceAccessManager.sol

```diff
@@ -22,201 +25,342 @@ contract InstanceAccessManager is
+            revert IAccess.ErrorIAccessTargetTypeInvalid(nameShort, _targetInfo[target].ttype);
+        }
+        // TODO isLocked is redundant but makes getTargetInfo() faster
+        _targetInfo[target].isLocked = locked;
         _accessManager.setTargetClosed(target, locked);
     }
 
-    function targetExists(address target) public view returns (bool exists) {
-        return _target[target].createdAt.gtz();
+    // allowed combinations of roles and targets:
+    //1) set core role for core target 
+    //2) set gif role for gif target  
+    //3) set custom role for gif target
+    //4) set custom role for custom target
+
+    // ADMIN_ROLE if used only during initialization, works with:
+    //      any roles for any targets
+    // INSTANCE_SERVICE_ROLE if used not only during initilization, works with:
+    //      core roles for core targets
+    //      gif roles for gif targets
+    // TODO taget admin example: 
+    // 1) INSTANCE_OWNER_ROLE is admin role for component target
+    // 2) ADMIN_ROLE is admin role for core targets
+    function setTargetFunctionRole(
+        string memory targetName,
+        bytes4[] calldata selectors,
+        RoleId roleId
+    ) 
+        public 
+        virtual 
+        restricted()
+    {
+        ShortString nameShort = ShortStrings.toShortString(targetName);
+        address target = _targetAddressForName[nameShort];
+
+        // not custom target
+        if(_targetInfo[target].ttype == IAccess.Type.Custom) {
+            revert IAccess.ErrorIAccessTargetTypeInvalid(nameShort, _targetInfo[target].ttype);
+        }
+
+        // not custom role
+        if(_roleInfo[roleId].rtype == IAccess.Type.Custom) {
+            revert IAccess.ErrorIAccessRoleTypeInvalid(roleId, _roleInfo[roleId].rtype);
+        }
+
+        _setTargetFunctionRole(target, nameShort, selectors, roleId);
     }
 
-    //--- internal view/pure functions --------------------------------------//
+    // INSTANCE_OWNER_ROLE
+    // custom role for gif target -> instance owner can mess with gif target (component) -> e.g. set custom role for function intendent to work with gif role
+    // custom role for custom target
+    function setTargetFunctionCustomRole(
+        string memory targetName,
+        bytes4[] calldata selectors,
+        RoleId roleId
+    ) 
+        public 
+        virtual 
+        restricted() 
+    {
+        ShortString nameShort = ShortStrings.toShortString(targetName);
+        address target = _targetAddressForName[nameShort];
+
+        // not core target
+        if(_targetInfo[target].ttype == IAccess.Type.Core) {
+            revert IAccess.ErrorIAccessTargetTypeInvalid(nameShort, _targetInfo[target].ttype);
+        }
 
-    function _createRole(RoleId roleId, string memory name, bool isCustom, bool validateParameters) internal {
-        if (validateParameters) {
-            _validateRoleParameters(roleId, name, isCustom);
+        // target belongs to instance owned by caller
```

Comment:
target might a contract that is not a component. Only validate if its a component. If its not a component, then just make sure its a registered target.

Reply:
moved this check to creatCustomTarget(): to become custom target address MUST not be registered
Assume that custom target can not be registered after creation